File: go/plugins/middleware/exp/artifacts_test.go
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

package exp

import (
	"context"
	"strings"
	"testing"

	"github.com/firebase/genkit/go/ai"
	aix "github.com/firebase/genkit/go/ai/exp"
	"github.com/firebase/genkit/go/genkit"
	genkitx "github.com/firebase/genkit/go/genkit/exp"
)

func TestArtifactsReadonlyOmitsWriteTool(t *testing.T) {
	hooks, err := (&Artifacts{Readonly: true}).New(ctx)
	if err != nil {
		t.Fatal(err)
	}
	if len(hooks.Tools) != 1 || hooks.Tools[0].Name() != "read_artifact" {
		t.Fatalf("readonly should expose only read_artifact, got %v", toolNames(hooks.Tools))
	}

	hooks, err = (&Artifacts{}).New(ctx)
	if err != nil {
		t.Fatal(err)
	}
	names := toolNames(hooks.Tools)
	if len(names) != 2 || !contains(names, "read_artifact") || !contains(names, "write_artifact") {
		t.Fatalf("default should expose read_artifact and write_artifact, got %v", names)
	}
}

func TestArtifactsWriteThenRead(t *testing.T) {
	g := newTestGenkit(t)

	// The model writes an artifact, reads it back, then finishes.
	model := toolModel(t, g, "test/artifact-model", func(ctx context.Context, req *ai.ModelRequest, cb ai.ModelStreamCallback) (*ai.ModelResponse, error) {
		var wrote, read bool
		for _, m := range req.Messages {
			for _, p := range m.Content {
				if p.IsToolResponse() && p.ToolResponse != nil {
					switch p.ToolResponse.Name {
					case "write_artifact":
						wrote = true
					case "read_artifact":
						read = true
					}
				}
			}
		}
		switch {
		case read:
			return textResp(req, "done"), nil
		case wrote:
			return toolReqResp(req, &ai.ToolRequest{Name: "read_artifact", Input: map[string]any{"name": "report.md"}}), nil
		default:
			return toolReqResp(req, &ai.ToolRequest{Name: "write_artifact", Input: map[string]any{"name": "report.md", "content": "hello world"}}), nil
		}
	})

	builder := genkitx.DefineAgent[any](g, "builder",
		aix.InlinePrompt{ai.WithModel(model), ai.WithSystem("be a builder"), ai.WithUse(&Artifacts{})},
	)

	out, err := builder.RunText(ctx, "make a report")
	if err != nil {
		t.Fatal(err)
	}

	// read_artifact returned the content written by write_artifact.
	reads := toolOutputs(statemessages(out), "read_artifact")
	if len(reads) == 0 {
		t.Fatal("no read_artifact tool response found")
	}
	read := decodeToolOutput[readArtifactOutput](t, reads[0])
	if !read.Found || read.Content != "hello world" {
		t.Errorf("read_artifact = %+v, want found with content %q", read, "hello world")
	}

	// The artifact persisted on the session.
	if !hasArtifactNamed(out.Artifacts, "report.md") {
		t.Errorf("expected artifact %q on session; got %v", "report.md", artifactNames(out.Artifacts))
	}
}

func TestArtifactsSystemPromptListing(t *testing.T) {
	g := newTestGenkit(t)

	var captured []*ai.Message
	capture := toolModel(t, g, "test/capture", func(ctx context.Context, req *ai.ModelRequest, cb ai.ModelStreamCallback) (*ai.ModelResponse, error) {
		captured = req.Messages
		return textResp(req, "ok"), nil
	})

	// A custom agent seeds an artifact, then generates with the Artifacts
	// middleware so the listing reflects the seeded artifact.
	lister := genkitx.DefineCustomAgent[any](g, "lister",
		func(ctx context.Context, resp aix.Responder, sess *aix.SessionRunner[any]) (*aix.AgentResult, error) {
			err := sess.Run(ctx, func(ctx context.Context, input *aix.AgentInput) (*aix.TurnResult, error) {
				resp.SendArtifact(&aix.Artifact{
					Name:  "notes.md",
					Parts: []*ai.Part{ai.NewTextPart("some notes here")},
				})
				if _, err := genkit.Generate(ctx, g,
					ai.WithModel(capture),
					ai.WithMessages(input.Message),
					ai.WithUse(&Artifacts{}),
				); err != nil {
					return nil, err
				}
				return &aix.TurnResult{FinishReason: aix.AgentFinishReasonStop}, nil
			})
			if err != nil {
				return nil, err
			}
			return &aix.AgentResult{Message: ai.NewModelTextMessage("listed")}, nil
		},
	)

	if _, err := lister.RunText(ctx, "what artifacts are there?"); err != nil {
		t.Fatal(err)
	}

	sys := findSystem(captured)
	if sys == nil {
		t.Fatalf("expected a system message; got %v", captured)
	}
	text := systemText(sys)
	if !strings.Contains(text, "<artifacts>") || !strings.Contains(text, "notes.md") {
		t.Errorf("system prompt missing artifact listing for notes.md; got:\n%s", text)
	}
}

func TestArtifactsNoSession(t *testing.T) {
	g := newTestGenkit(t)

	// With a plain Generate call there is no agent session.
	model := toolModel(t, g, "test/no-session", func(ctx context.Context, req *ai.ModelRequest, cb ai.ModelStreamCallback) (*ai.ModelResponse, error) {
		for _, m := range req.Messages {
			for _, p := range m.Content {
				if p.IsToolResponse() {
					return textResp(req, "done"), nil
				}
			}
		}
		return toolReqResp(req, &ai.ToolRequest{Name: "read_artifact", Input: map[string]any{"name": "x"}}), nil
	})

	resp, err := genkit.Generate(ctx, g, ai.WithModel(model), ai.WithPrompt("read x"), ai.WithUse(&Artifacts{}))
	if err != nil {
		t.Fatal(err)
	}
	reads := toolOutputs(resp.History(), "read_artifact")
	if len(reads) == 0 {
		t.Fatal("no read_artifact tool response found")
	}
	read := decodeToolOutput[readArtifactOutput](t, reads[0])
	if read.Found || !strings.Contains(read.Content, "no active session") {
		t.Errorf("expected a no-active-session result, got %+v", read)
	}
}

func toolNames(tools []ai.Tool) []string {
	names := make([]string, 0, len(tools))
	for _, tl := range tools {
		names = append(names, tl.Name())
	}
	return names
}

func contains(ss []string, s string) bool {
	for _, x := range ss {
		if x == s {
			return true
		}
	}
	return false
}

File: py/packages/genkit-middleware/tests/artifacts_test.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Tests for Artifacts middleware."""

from __future__ import annotations

import pytest
from genkit_middleware import Artifacts
from genkit_middleware._artifacts import (
    ARTIFACTS_LISTING_MARKER,
    build_artifact_listing,
    extract_artifact_text,
)

from genkit import ModelResponse
from genkit._ai._agents._session import Session, run_with_session
from genkit._core._model import GenerateActionOptions
from genkit._core._typing import Artifact, Part, Role, SessionState, TextPart
from genkit.middleware import GenerateHookParams, GenerateMiddlewareContext


def _make_params(options: GenerateActionOptions | None = None) -> GenerateHookParams:
    opts = options or GenerateActionOptions(messages=[])
    return GenerateHookParams(
        options=opts,
        iteration=0,
    )


def _listing_parts(messages) -> list[TextPart]:
    parts: list[TextPart] = []
    for msg in messages:
        if msg.role != Role.SYSTEM:
            continue
        for part in msg.content:
            root = part.root
            if isinstance(root, TextPart) and isinstance(root.metadata, dict):
                if root.metadata.get(ARTIFACTS_LISTING_MARKER):
                    parts.append(root)
    return parts


def test_build_artifact_listing_empty() -> None:
    listing = build_artifact_listing([])
    assert 'No artifacts are currently available' in listing
    assert listing.startswith('<artifacts>')


def test_extract_artifact_text() -> None:
    art = Artifact(name='a.txt', parts=[Part(TextPart(text='line1')), Part(TextPart(text='line2'))])
    assert extract_artifact_text(art) == 'line1\nline2'


@pytest.mark.asyncio
async def test_write_artifact_uses_current_session(ctx: GenerateMiddlewareContext) -> None:
    mw = Artifacts()
    session = Session(SessionState())

    async def check() -> None:
        tools = {t.name: t for t in mw.tools(ctx)}
        assert set(tools) == {'read_artifact', 'write_artifact'}

        write = tools['write_artifact']
        result = await write.run(input={'name': 'poem.txt', 'content': 'roses are red'})
        assert result.response.output['status'] == 'Artifact "poem.txt" saved successfully.'
        arts = await session.get_artifacts()
        assert len(arts) == 1
        assert arts[0].name == 'poem.txt'
        assert arts[0].parts[0].root.text == 'roses are red'

    await run_with_session(session=session, coro=check())


@pytest.mark.asyncio
async def test_read_artifact_returns_found(ctx: GenerateMiddlewareContext) -> None:
    mw = Artifacts()
    session = Session(SessionState())
    await session.add_artifacts([Artifact(name='notes.txt', parts=[Part(TextPart(text='hello'))])])

    async def check() -> None:
        read = next(t for t in mw.tools(ctx) if t.name == 'read_artifact')

        result = await read.run(input={'name': 'notes.txt'})
        assert result.response.output['name'] == 'notes.txt'
        assert result.response.output['content'] == 'hello'
        assert result.response.output['found'] is True

    await run_with_session(session=session, coro=check())


@pytest.mark.asyncio
async def test_read_artifact_without_session(ctx: GenerateMiddlewareContext) -> None:
    mw = Artifacts()
    read = next(t for t in mw.tools(ctx) if t.name == 'read_artifact')
    result = await read.run(input={'name': 'missing.txt'})
    assert result.response.output['name'] == 'missing.txt'
    assert 'no active agent session' in result.response.output['content'].lower()
    assert result.response.output['found'] is False


@pytest.mark.asyncio
async def test_readonly_excludes_write_tool(ctx: GenerateMiddlewareContext) -> None:
    mw = Artifacts(readonly=True)
    names = {t.name for t in mw.tools(ctx)}
    assert names == {'read_artifact'}


@pytest.mark.asyncio
async def test_wrap_generate_injects_listing(ctx: GenerateMiddlewareContext) -> None:
    mw = Artifacts()
    session = Session(
        SessionState(artifacts=[Artifact(name='poem.txt', parts=[Part(TextPart(text='abc'))])]),
    )

    captured: list[GenerateActionOptions] = []

    async def next_fn(params, _ctx):
        captured.append(params.options)
        return ModelResponse(message=None)

    async def check() -> None:
        await mw.wrap_generate(_make_params(), ctx, next_fn)

        assert len(captured) == 1
        system_msgs = [m for m in captured[0].messages if m.role == Role.SYSTEM]
        assert len(system_msgs) == 1
        listing_parts = [
            p
            for p in system_msgs[0].content
            if isinstance(p.root, TextPart)
            and isinstance(p.root.metadata, dict)
            and p.root.metadata.get(ARTIFACTS_LISTING_MARKER)
        ]
        assert len(listing_parts) == 1
        assert 'poem.txt' in (listing_parts[0].root.text or '')
        assert '(3 chars)' in (listing_parts[0].root.text or '')

    await run_with_session(session=session, coro=check())


@pytest.mark.asyncio
async def test_wrap_generate_refreshes_listing(ctx: GenerateMiddlewareContext) -> None:
    mw = Artifacts()
    session = Session(SessionState())
    envelope = GenerateActionOptions(messages=[])

    seen: list[str] = []

    async def next_fn(params, _ctx):
        for part in _listing_parts(params.options.messages):
            seen.append(part.text or '')
        return ModelResponse(message=None)

    async def check() -> None:
        await mw.wrap_generate(_make_params(envelope), ctx, next_fn)
        await session.add_artifacts([Artifact(name='b.txt', parts=[Part(TextPart(text='x'))])])
        await mw.wrap_generate(_make_params(envelope), ctx, next_fn)

        assert len(seen) == 2
        assert 'No artifacts are currently available' in seen[0]
        assert 'b.txt' in seen[1]
        assert len(_listing_parts(envelope.messages)) == 0

    await run_with_session(session=session, coro=check())


@pytest.mark.asyncio
async def test_wrap_generate_does_not_mutate_envelope(ctx: GenerateMiddlewareContext) -> None:
    mw = Artifacts()
    envelope = GenerateActionOptions(messages=[])
    session = Session(
        SessionState(artifacts=[Artifact(name='a.txt', parts=[Part(TextPart(text='hi'))])]),
    )

    captured_request: list[GenerateActionOptions] = []

    async def next_fn(params, _ctx):
        captured_request.append(params.options)
        return ModelResponse(message=None)

    async def check() -> None:
        await mw.wrap_generate(_make_params(envelope), ctx, next_fn)

        assert len(_listing_parts(envelope.messages)) == 0
        assert len(_listing_parts(captured_request[0].messages)) == 1
        assert 'a.txt' in _listing_parts(captured_request[0].messages)[0].text

    await run_with_session(session=session, coro=check())
